### PR TITLE
Add `static` folder for static files and pre-loaded stylesheets (using new utility)

### PR DESCRIPTION
**Pull request by edan-bainglass**
This PR adds a `static` folder to the widgets base package. At the moment, the folder holds stylesheets that are loaded (using a new utility function) on any import from the widgets base package.

### CODE REVIEW

File: aiidalab_widgets_base/__init__.py
Comment (danielhollas) [inline aiidalab_widgets_base/__init__.py:51]: To improve readability slightly, I'd move this to the top before the loading of the AiiDA profile.

File: aiidalab_widgets_base/utils/loaders.py
Comment (danielhollas) [inline aiidalab_widgets_base/utils/loaders.py:6]: We currently don't check types so I'd remove it. Or alternatively, type the other parameter and return parameter as well.

File: aiidalab_widgets_base/static/styles/README.md
Comment (danielhollas) [inline aiidalab_widgets_base/static/styles/README.md:10]: Does this mean that compiled CSS files should be checked in the repo? If so, we need a pre-commit hook that will check that the files are up-to-date. Alternatively, we need to create a hook that is called when the package is installed.

File: aiidalab_widgets_base/static/styles/README.md
Comment (danielhollas) [inline aiidalab_widgets_base/static/styles/README.md:13]: I think we should have a simple bash script that will invoke the correct invocation of the compiler. Otherwise it's not immediately clear what needs to be done.  (this could then be used as a pre-commit hook, as noted above).

**Review by danielhollas — CHANGES_REQUESTED** [review]
Hi @edan-bainglass,

thanks so much on working on this, I am really excited to see this land.

My main concern here is about SCSS, and is twofold.
- It introduces additional complexity of extra compilation step (see my other comments)
- I guess that most of us are not so proficient with CSS, and SCSS is yet another cognitive overhead. Especially as we just start using CSS directly, my feeling is that we should start simple.

If you feel strongly about it, could you please open an issue (or topic on Discourse), and explain what are the advantages of SCSS (and what is the additional complexity) so we can discuss it a bit before committing?

File: aiidalab_widgets_base/__init__.py
Comment (edan-bainglass) [reply aiidalab_widgets_base/__init__.py:51]: I think it had to be after. Otherwise, I got a warning message about not importing from the widgets base prior to loading the profile

File: aiidalab_widgets_base/utils/loaders.py
Comment (edan-bainglass) [reply aiidalab_widgets_base/utils/loaders.py:6]: The other parameter is typed via its default (str). The return type is None, so not strictly required. Do you feel differently about this? I'm not against it. In principle, I know packages such as pydantic enforce typing even in cases where static analysis tools can assert it.

File: aiidalab_widgets_base/static/styles/README.md
Comment (edan-bainglass) [reply aiidalab_widgets_base/static/styles/README.md:10]: I gotta be honest, I'm not all that fond of the way I did it. But I couldn't come up with an alternative. It would be great to implement something more flexible, where one can use both CSS and SCSS. Let's chat on Discourse.

File: aiidalab_widgets_base/static/styles/README.md
Comment (edan-bainglass) [reply aiidalab_widgets_base/static/styles/README.md:13]: See my last comment

**Comment by edan-bainglass** [comment]
@danielhollas thanks for your comments. Let's discuss further on [Discourse](https://aiida.discourse.group/t/use-of-external-stylesheets-css-or-even-scss/434).

File: aiidalab_widgets_base/__init__.py
Comment (danielhollas) [reply aiidalab_widgets_base/__init__.py:51]: ah, of course, makes sense.

**Comment by edan-bainglass** [comment]
@danielhollas could you please take a look at why the browser tests are failing? I'm not sure I understand what's happening here.

**Comment by danielhollas** [comment]
@edan-bainglass The same failures are seen on the main branch so it is not related to this PR.

**Comment by edan-bainglass** [comment]
@danielhollas I see. Okay then. Please review 🙏🏻 I have a few PRs built on top of this one both here and in the QE app. I think all is well other than the tests. I've removed the SCSS part.

File: aiidalab_widgets_base/utils/loaders.py
Comment (edan-bainglass) [reply aiidalab_widgets_base/utils/loaders.py]: I see. I've added the explicit `str` type :)

File: aiidalab_widgets_base/utils/loaders.py
Comment (danielhollas) [inline aiidalab_widgets_base/utils/loaders.py:27]: Maybe a bit simpler
```suggestion
    if filename:
        filenames = [filename]
    else:
        filenames = [
            path.name
            for path in root.iterdir()
            if path.is_file() and path.name.endswith(".css")
        ]
```

File: aiidalab_widgets_base/utils/loaders.py
Comment (danielhollas) [inline aiidalab_widgets_base/utils/loaders.py:30]: nit: This for loop var shadows the function parameter. Not a huge deal but I am surprised that ruff does not complain about it.

```suggestion
    for f in filenames:
        stylesheet = (root / f).read_text()
```

**Review by danielhollas — CHANGES_REQUESTED** [review]
Just some small nits.

My main thing is that it would be nice to demonstrate this functionality to prove that is working by including some actual CSS.

Perhaps a nice candidate would be this code that defines CSS inline for tables in DictViewer.

https://github.com/aiidalab/aiidalab-widgets-base/blob/de5f0cc23e1e058b46922a2203dd75b292b0f13e/aiidalab_widgets_base/viewers.py#L104-L113

File: aiidalab_widgets_base/__init__.py
Comment (edan-bainglass) [reply aiidalab_widgets_base/__init__.py:46]: Done!

File: aiidalab_widgets_base/utils/loaders.py
Comment (edan-bainglass) [reply aiidalab_widgets_base/utils/loaders.py:30]: Indeed surprising. Done!

File: aiidalab_widgets_base/utils/loaders.py
Comment (edan-bainglass) [reply aiidalab_widgets_base/utils/loaders.py:27]: Done!

File: aiidalab_widgets_base/__init__.py
Comment (superstar54) [inline aiidalab_widgets_base/__init__.py:14]: Could you move this to the `utils/__init__.py`?

File: aiidalab_widgets_base/__init__.py
Comment (superstar54) [inline aiidalab_widgets_base/__init__.py:48]: Will the next major release fix this?

File: aiidalab_widgets_base/utils/loaders.py
Comment (superstar54) [inline aiidalab_widgets_base/utils/loaders.py:6]: I suggest changing `package` to `module`, as it accepts a module.

File: tests/test_loaders.py
Comment (superstar54) [inline tests/test_loaders.py:8]: The test here will not work because it is not running inside a notebook. To test whether the CSS works, you need to add a test in `test_notebooks/test_notebooks.py.`

**Review by superstar54 — CHANGES_REQUESTED** [review]
Hi @edan-bainglass , thansk for the work.

- It would be great that you add this in the docs to tell the developer how to use it in their app.
- You need add your test in the `test_notebooks`.

File: aiidalab_widgets_base/__init__.py
Comment (danielhollas) [reply aiidalab_widgets_base/__init__.py:14]: This needs to stay here, this code was here before and was just moved to a function.

File: aiidalab_widgets_base/__init__.py
Comment (danielhollas) [reply aiidalab_widgets_base/__init__.py:48]: This is unrelated to this PR. Indeed we plan to remove this in next major release most likely.

File: aiidalab_widgets_base/__init__.py
Comment (superstar54) [reply aiidalab_widgets_base/__init__.py:14]: Ah, you are right!

File: aiidalab_widgets_base/__init__.py
Comment (superstar54) [reply aiidalab_widgets_base/__init__.py:48]: Ok.

File: aiidalab_widgets_base/utils/loaders.py
Comment (danielhollas) [reply aiidalab_widgets_base/utils/loaders.py:6]: I think the name is correct, it accepts a package (i.e. a folder), not a module (which would be an individual python file).

File: tests/test_loaders.py
Comment (edan-bainglass) [reply tests/test_loaders.py:8]: Are you saying to add a test to test_notebooks.py, or to move this test to test_notebooks.py? Also, could you explain why exactly it doesn't work?

File: aiidalab_widgets_base/utils/loaders.py
Comment (superstar54) [reply aiidalab_widgets_base/utils/loaders.py:6]: `package` is more like a Python pacage, like the `aiidalab_widgets_base`. From the example, 

```python
    from .static import styles
    from .utils.loaders import load_css_stylesheet

    load_css_stylesheet(package=styles)
```
It passes the module `styles` into it.  'aiidalab_widgets_base.static.styles' is a module, not a package.

File: tests/test_loaders.py
Comment (superstar54) [reply tests/test_loaders.py:8]: Yes, adding a new test in the test_notebooks.py, and check if the CSS is indeed loaded in the selenium_driver.

> Also, could you explain why exactly it doesn't work?

the `test.py` file is not running in the notebook, thus the `display(Javascript...` code will not run. You can add some wrong code inside it, and the test will not throw error.

**Comment by edan-bainglass** [comment]
> Just some small nits.
> 
> My main thing is that it would be nice to demonstrate this functionality to prove that is working by including some actual CSS.
> 
> Perhaps a nice candidate would be this code that defines CSS inline for tables in DictViewer.
> 
> https://github.com/aiidalab/aiidalab-widgets-base/blob/de5f0cc23e1e058b46922a2203dd75b292b0f13e/aiidalab_widgets_base/viewers.py#L104-L113
> 
> EDIT: I know that you have follow-up PRs that do this so feel free to ignore this.

@danielhollas yeah, maybe an example is good just for this commit. Since you've suggested one from the code base, it's a great test case. Matter of fact, I could use a test case for a notebook test, as recommended by @superstar54. So to clarify, the bit of code you mentioned (`df` styles) can be removed once I add it via css, correct?

File: aiidalab_widgets_base/utils/loaders.py
Comment (danielhollas) [reply aiidalab_widgets_base/utils/loaders.py:6]: I don't think that's correct, citing e.g. from [this glossary](https://the-hitchhikers-guide-to-packaging.readthedocs.io/en/latest/glossary.html)

> **Module**
>     A python source code file (ex. mymodule.py), most often found in a package (mypackage/mymodule.py). You can import a module: import mymodule, or import mypackage.mymodule if it lives in mypackage.
>
> **Package** 
>     A directory containing an __init__.py file (ex. mypackage/__init__.py), and also usually containing modules (possibly along with other packages). You can import a package: import mypackage
>     A package should not be confused with a compressed archive file used to install Python code.

`aiidalab_widgets_base` is a package, containing `static` subpackage which itself contains `styles` subpackage. After all, the type of this variable *is* called `Package`.

File: tests/test_loaders.py
Comment (superstar54) [reply tests/test_loaders.py:8]: For the test in selenium, for example, you create a vbox, and attach a class, then in your testing CSS  file, you change its color, and in selenium_driver, you try to find the vbox, and verify the color.

Comment (danielhollas) [reply tests/test_loaders.py:8]: The test here is fine, I think but I agree it would be nice to have *an additional` test in `test_notebooks` that runs an actual browser via selenium that would prove that the CSS is indeed loaded and applied to a given element (I don't actually know if selenium allows that but I would expect so).

Comment (edan-bainglass) [reply tests/test_loaders.py:8]: Will add. Thanks

**Comment by edan-bainglass** [comment]
@danielhollas proceeding with a made-up class for testing purposes only, as to not interfere with the code base for now.

File: tests/test_loaders.py
Comment (edan-bainglass) [reply tests/test_loaders.py:8]: @superstar54 I'm not familiar with selenium. I see that you have to set a driver which wraps on an existing notebook. I then do `driver.find_element`, but does that only check THAT notebook? If so, which notebook do I use?

Comment (superstar54) [reply tests/test_loaders.py:8]: I think you can create a new notebook (e.g., `test_css.ipynb`), and in one cell you add
```python
label = ipw.Label("Testing")
label.add_class("for-testing-purposes")
display(label)
```

in your test
```python
def test_load_css_stylesheet(selenium_driver):
    """Test `load_css_stylesheet` function."""
    driver = selenium_driver("notebooks/test_css.ipynb")
    driver.find_element(By.CLASS_NAME, "for-testing-purposes")
```
you may investigate how to check the style of the element, I don't know the how to do this exactly.

Comment (edan-bainglass) [reply tests/test_loaders.py:8]: Sure, if you guys don't mind another notebook just for testing. Though maybe just call it `test.ipynb`? You never know what other tests might need a dedicated notebook🤷🏻‍♂️

Comment (danielhollas) [reply tests/test_loaders.py:8]: I feel like there is a bit of misunderstanding here. The purpose of the notebook test is not to test the load_css_stylesheet on its own. It should be mainly test that the CSS style is actually applied. 

If you added the CSS code for the DictViewer, you could just extend the existing DictViewer test. As @superstar54 said, you'll need to investigate if selenium allows you to actually test CSS properties. If not, then adding a notebook test does not give us much.

The test itself shouldn't be using `load_css_stylesheet`, it should be loaded implicitly when awb is imported.

File: aiidalab_widgets_base/utils/loaders.py
Comment (superstar54) [reply aiidalab_widgets_base/utils/loaders.py:6]: Thanks for the clarification! But it is still very weird to call, e.g., `aiida.orm`, a package.

File: tests/test_loaders.py
Comment (edan-bainglass) [reply tests/test_loaders.py:8]: Take a look at the notebook test I pushed. It does exactly that - it checks that a property of a class is present on a widget. However, it does use `load_css_stylesheet` in the test. This would be redundant if awb is imported, but is it imported anywhere? I only imported it in the test to use `load_css_stylesheet`. If I remove the call, where would awb be called?

Comment (superstar54) [reply tests/test_loaders.py:8]: Yes, in the test notebook, you need at least 
```python
import aiidalab_widgets_base
```

Comment (edan-bainglass) [reply tests/test_loaders.py:8]: Why?

Comment (superstar54) [reply tests/test_loaders.py:8]: I mean you only need to import `aiidalab_widgets_base`, since it will call `load_css_stylesheet` in the `__init__.py` file.

Comment (edan-bainglass) [reply tests/test_loaders.py:8]: I've loaded `aiidalab_widgets_base` in the `test.ipynb` notebook with a comment regarding its use, as it is not actually used explicitly.

Comment (edan-bainglass) [reply tests/test_loaders.py:8]: Okay, notebook test passes 👍

**Comment by edan-bainglass** [comment]
I think this is good now? Failed tests unrelated I believe.

**Review by superstar54 — APPROVED** [review]
LGTM! Thanks for the nice work!

File: aiidalab_widgets_base/static/styles/test.css
Comment (edan-bainglass) [reply aiidalab_widgets_base/static/styles/test.css:1]: Sure, but then I'd have to explicitly load it in the test notebook.

Comment (edan-bainglass) [reply aiidalab_widgets_base/static/styles/test.css:1]: So the idea is to test that the loader works. But test_notebooks is not a package, so how would I load `test.css` from there?

Comment (edan-bainglass) [reply aiidalab_widgets_base/static/styles/test.css]: I rather not mess with an internal component, lest I break it 😬 The style of `DictViewer` is assigned to `self.value`. `value` is an observed property, so not touching it, nor do I have time to understand what is happening in this code 🤷🏻‍♂️. But you're welcome to contribute to this PR with the proposed change 🙂

For now, I've modified the CSS loader utility and used it in the test notebook. I think this is acceptable, but again, if you feel strongly about it, see above comment.

Comment (danielhollas) [reply aiidalab_widgets_base/static/styles/test.css]: Okay, fair enough, sorry for the back-and-forth. However, I really don't like the changes to the loader function. Feels like the complexity greatly increase only for the purposes of testing. :-(

File: aiidalab_widgets_base/utils/loaders.py
Comment (danielhollas) [inline aiidalab_widgets_base/utils/loaders.py:13]: Having both `package` and `css_path` makes this interface suuuuper messy.

How about we instead only have a single parameter of type `Path` that would be a directory with the css files. (maybe `css_dir` would be a better name, since `css_path` does not make it clear that it is not a path to a single CSS filename. 

It would then be the callers responsibility to convert Package to Path if needed.

@edan-bainglass does that make sense?

Comment (edan-bainglass) [reply aiidalab_widgets_base/utils/loaders.py:13]: Sure

Comment (edan-bainglass) [reply aiidalab_widgets_base/utils/loaders.py:13]: Hmm, on second thought, the only reason I introduced another use case for the API is due to the relocation of the `test.css` stylesheet outside of the package. So, as stated in the comment regarding `DictViewer`, you are welcome to modify the test to simply look at custom styles for the `DictViewer`, at which point, most of the complexity can be removed.

Comment (danielhollas) [reply aiidalab_widgets_base/utils/loaders.py:13]: I will not have time to dig into into myself at this point, sorry. I think my suggestion makes sense regardless of the motivation imo. To me it makes more sense if the function accepts a Path to CSS dir instead of Package, it makes it more versatile, and it's users are free to decide where they want to put the css files (i.e. not necessarily  inside the module).

Comment (edan-bainglass) [reply aiidalab_widgets_base/utils/loaders.py:13]: Understood. I will take a look at implementing it. Thanks 🙂

Comment (danielhollas) [reply aiidalab_widgets_base/utils/loaders.py:13]: Thanks for bearing with me. Once we merge this I'll release a new alpha version. We can still iterate on it until the final `2.3.0` release.

**Comment by edan-bainglass** [comment]
I'll fix the tests in the morning 😮‍💨

File: aiidalab_widgets_base/utils/loaders.py
Comment (danielhollas) [inline aiidalab_widgets_base/utils/loaders.py:8]: Thank you, this looks better! 

...now that we're passing a path, I am wondering, it would seem more natural to have a single parameter, that can be either a dir or a file. The split to two parameters for dir and file doesn't seem intuitive, and I don't see any advantage of doing it that way. 

(Sorry, I will shut up eventually : 😅) But perhaps good to iterate on this interface now rather to make it as intuitive as possible).

Comment (edan-bainglass) [reply aiidalab_widgets_base/utils/loaders.py:8]: I'd prefer to keep the dir and filename separate. The reason for it is clear in the logic of the function.

Comment (danielhollas) [reply aiidalab_widgets_base/utils/loaders.py:8]: Not sure if the function logic should define its interface. From a users perspective, passing a single path seems such simpler to think about (also it's not great that the two parameters have different type, another thing that can trip the user).

@superstar WDYT? 

(otherwise LGTM, thanks!)

Comment (danielhollas) [reply aiidalab_widgets_base/utils/loaders.py:8]: Sorry, I guess I was replying to the older version of your comment. 

Yes, I would expect the the function can internally use Path.is_dir() to determine what to do.

Otherwise I am happy with the changes.

Comment (edan-bainglass) [reply aiidalab_widgets_base/utils/loaders.py:8]: Your comments regarding the API are valid. In general, the user comes first (reasonably). I'll adjust the code in the morning. Thanks.

Comment (edan-bainglass) [reply aiidalab_widgets_base/utils/loaders.py]: API now takes `css_path` of type `Path` or `str` (the latter converted to `Path`).

Comment (danielhollas) [reply aiidalab_widgets_base/utils/loaders.py]: Looks great, thank you!

**Review by danielhollas — APPROVED** [review]
LGTM. :rocket: Thanks for bearing with me, I am super happy with the end result.

**Comment by edan-bainglass** [comment]
@danielhollas what now? You merge? Also, docs build failed?

**Comment by danielhollas** [comment]
@edan-bainglass if you have permissions to merge go ahead. If not let me know and I'll merge.

The doc failure is `

```
checking consistency... /home/docs/checkouts/readthedocs.org/user_builds/aiidalab-widgets-base/checkouts/624/docs/source/widget-list/notebooks/test_misc.ipynb: WARNING: document isn't included in any toctree
```

Please investigate how get rid of this warning or ignore it. I suspect the easiest solution is to move the notebook into the `notebook_tests` directory since it is not supposed to be in the docs.

**Comment by edan-bainglass** [comment]
@danielhollas @superstar54 committed a bit of documentation. Please give it a looksee 🙏 Also, @danielhollas, the notebook tests are still failing 😔

**Comment by danielhollas** [comment]
The tests are passing on another PR.

**Comment by edan-bainglass** [comment]
Found some oversights in my code. Fixed it. Still not passing the tests. But its not just my test. @danielhollas @superstar54 could you please have a look?

**Comment by danielhollas** [comment]
@edan-bainglass have you tried running the tests locally? That should make it clearer why they are failing.

**Comment by edan-bainglass** [comment]
I couldn't figure out how to run selenium tests locally. I'll give it another go.

**Comment by danielhollas** [comment]
@edan-bainglass congrats on merging this. :tada: I'll make a new alpha release that you can then use on PRs on QEApp. 

Could you please write here what you did to fix the tests and if / how you were able to test notebooks locally? I'll then add it to README for contributors. Thanks!

**Comment by edan-bainglass** [comment]
@danielhollas thanks 🙏 Though I'm really not sure what actually fixed it. I was unable to run selenium tests locally in the container. I just applied the ol' step-by-step approach, walking back my commits, pushing chunks at a time and observing the tests. The CSS loading on import using the new utility passed the tests. Adding the notebook test in the notebooks folder with the test CSS as part of the module passed the tests. Then finally migrating the test notebook and CSS to the notebook tests folder... also passed the tests 🤷‍♂️ So in the end, I actually changed nothing from the state at which tests were failing. This was very confusing in general, as it was tests unrelated to my own that were failing before. Something to do with the selenium driver itself not finding the elements stated by the tests. According to [this article](https://reflect.run/articles/everything-you-need-to-know-about-nosuchelementexception-in-selenium/), "[NoSuchElementException](https://www.selenium.dev/selenium/docs/api/java/org/openqa/selenium/NoSuchElementException.html) is one of the most common exceptions in Selenium WebDriver..."

Of course, happy to get another set of eyes on it. Maybe I missed something.

**Pull request merged**
